### PR TITLE
Add ComparePath func

### PR DESCRIPTION
**Pull request by DanG100**
i also broke the circular dependency by renaming the package in this file to util_test

### CODE REVIEW

File: util/gnmi.go [hunk 1]
```diff
@@ -255,3 +255,101 @@ func JoinPaths(prefix, suffix *gpb.Path) (*gpb.Path, error) {
...
+		relation = Subset
+		shortestLen = len(b.Elem)
+	} else if len(a.Elem) < len(b.Elem) {
+		relation = Superset
+	}
+
+	for i := 0; i < shortestLen; i++ {
+		if a.Elem[i].Name != b.Elem[i].Name {
+			return Disjoint, nil
+		}
```
Comment (wenovus) [inline util/gnmi.go:294]: Can put this in comparePathElem and then add a switch case.

File: util/gnmi.go [hunk 2]
```diff
@@ -255,3 +255,101 @@ func JoinPaths(prefix, suffix *gpb.Path) (*gpb.Path, error) {
...
+// error: not two keys are both subset and superset.
+func comparePathElem(a, b *gpb.PathElem) (CompareRelation, error) {
+	setRelation := Equal
+	for k, aVal := range a.Key {
+		bVal, ok := b.Key[k]
+		switch {
+		case aVal == bVal: // Values are equal.
+			continue
+		case aVal == "*" && !ok: // Values are equal, matching explicit wildcard with implicit one.
+			continue
```
Comment (wenovus) [inline util/gnmi.go:324]: merge these two cases
`case aVal == bVal, aVal == "*" && !ok`

File: util/gnmi.go [hunk 3]
```diff
@@ -255,3 +255,101 @@ func JoinPaths(prefix, suffix *gpb.Path) (*gpb.Path, error) {
...
+
+	return relation, nil
+}
+
+// comparePathElem compare two path elements a, b and returns:
+// subset: if every definite key in a is wildcard in b.
+// superset: if every wildcard key in b is non-wildcard in b.
+// error: not two keys are both subset and superset.
+func comparePathElem(a, b *gpb.PathElem) (CompareRelation, error) {
+	setRelation := Equal
```
Comment (wenovus) [inline util/gnmi.go:317]: // start by assuming a perfect match. Then relax this property as we scan through the key values.

File: util/gnmi.go [hunk 4]
```diff
@@ -255,3 +255,101 @@ func JoinPaths(prefix, suffix *gpb.Path) (*gpb.Path, error) {
...
+				return Disjoint, fmt.Errorf("path elem %v is not consistently a subset/superset of %v", a, b)
+			}
+			setRelation = Superset
+		case bVal == "*" || !ok: // Values not equal, a value is subset of b value.
+			if setRelation == Superset {
+				return Disjoint, fmt.Errorf("path %v is not consistently a subset/superset of %v", a, b)
+			}
+			setRelation = Subset
+		default: // Values not equal, but of the same size.
+			setRelation = Disjoint
```
Comment (wenovus) [inline util/gnmi.go:336]: can just return here?

File: util/gnmi.go [hunk 5]
```diff
@@ -255,3 +255,101 @@ func JoinPaths(prefix, suffix *gpb.Path) (*gpb.Path, error) {
...
+		}
+	}
+	for k, bVal := range b.Key {
+		_, ok := a.Key[k]
+		switch {
+		case ok: // Already visited all the keys in a.
+			continue
+		case bVal == "*": // Values are equal, matching implicit wildcard.
+			continue
```
Comment (wenovus) [inline util/gnmi.go:345]: merge these two cases.

**Review by wenovus — CHANGES_REQUESTED** [review]
More nitpicks.

File: util/gnmi.go [hunk 6]
```diff
@@ -255,3 +255,100 @@ func JoinPaths(prefix, suffix *gpb.Path) (*gpb.Path, error) {
...
+	} else if len(a.Elem) < len(b.Elem) {
+		relation = Superset
+	}
+
+	for i := 0; i < shortestLen; i++ {
+		elemRelation := comparePathElem(a.Elem[i], b.Elem[i])
+		switch {
+		case elemRelation == PartialIntersect:
+			return PartialIntersect
+		case (relation == Subset && elemRelation == Superset) || (relation == Superset && elemRelation == Subset):
```
Comment (wenovus) [inline util/gnmi.go:298]: nit: || -> ,
then can remove brackets

File: util/gnmi.go [hunk 7]
```diff
@@ -255,3 +255,100 @@ func JoinPaths(prefix, suffix *gpb.Path) (*gpb.Path, error) {
...
+			return PartialIntersect
+		case elemRelation == Subset || elemRelation == Superset:
+			relation = elemRelation
+		case elemRelation == Disjoint && relation == Equal:
+			relation = Disjoint
```
Comment (wenovus) [inline util/gnmi.go:303]: Why is `relation == Equal` a condition? If the current elem is disjoint then the whole path is also disjoint right? Doesn't matter what happened before.

**Review by wenovus — APPROVED** [review]

File: util/gnmi.go (hunk 7 above)
Comment (DanG100) [reply util/gnmi.go]: simplified even further

**Review by wenovus — APPROVED** [review]
Looks great, thanks

**Pull request merged**